R/add_bottom_borders.R

```diff
@@ -0,0 +1,196 @@
+#' and no borders in the second row.
+#'
+#' A border matrix:
+#' | 1 | 2 |
+#' | - | - |
+#' | 0 | 0 |
+#'
+#' The above border matrix puts one border under the first cell in the first row;
+#' and another border (separate from the first one) under the second cell
+#' in the first row. The second row stays without any borders.
+#'
+#' @param ht `huxtable` A huxtable object
+#' @param border_matrix (optional) `matrix` A matrix indicating where to add the bottom
+#' borders.
+#' @param transform_fns (optional) `list` of `function` A list of functions applied to the
+#' `border_matrix`. The functions have to accept two arguments:
+#' 1. The `huxtable`.
+#' 1. The `border_matrix`.
+#'
+#' The function have to return a matrix with the same dimensions as `huxtable`
+#' indicating where to put the borders. The functions in the list are applied
+#' sequentially to `border_matrix`.
+#'
+#' @return A huxtable with added borders.
+#'
+#' @examples
+#' border_matrix <- matrix(c(1, 1, 2, 0, 1, 1, 0, 0, 0), nrow = 3, ncol = 3)
+#' ht <- huxtable::as_huxtable(
+#'   data.frame(a = c(1, 2, 3), b = c("a", "b", "c"), c = c(TRUE, FALSE, TRUE))
+#' )
+#' # By default adds no borders
+#' add_bottom_borders(ht, border_matrix)
+#' #'
+#' # Adds spanning borders under cells with text in the second row
+#' add_bottom_borders(ht, transform_fns = list(spanning_borders(2)))
+#' # Adds spanning borders under cells with text in the second row and a border
+#' # under a cell in row 3 and column 3
+#' add_bottom_borders(ht, transform_fns = list(spanning_borders(2), single_border(3, 3)))
+#'
+#' @export
+add_bottom_borders <- function(ht, border_matrix = no_borders(ht), transform_fns = list()) {
+  if (is.null(border_matrix) && is.null(transform_fns)) {
+    return(ht)
+  } else if (is.null(border_matrix)) {
+    border_matrix <- no_borders(ht)
+  }
+  for (transform_fn in transform_fns) {
+    border_matrix <- transform_fn(ht, border_matrix)
+  }
+
+  internal_bottom_border <- r"{\brdrb\brdrs}"
+  solid_border <- huxtable::brdr(thickness = 0.9, style = "solid")
+  for (row in seq_len(nrow(border_matrix))) {
+    # The below enforces an "all or nothing" behaviour on using
+    # one of the border styles, because mixing them in the
+    # scope of a row gives bad results - the borders
+    # are misaligned and of slightly different width
+    should_use_internal_borders <- should_use_internal_borders(border_matrix[row, ])
+
+    for (col in seq_len(ncol(border_matrix))) {
+      if (border_matrix[row, col] != 0) {
+        if (should_use_internal_borders) {
+          ht <- huxtable::set_contents(
+            ht,
+            row,
+            col,
+            sprintf("%s %s", internal_bottom_border, ht[row, col])
+          )
+          if (col == ncol(border_matrix)) {
+            ht <- huxtable::set_right_padding(ht, row, col, 0)
+          }
+        } else {
+          ht <- huxtable::set_bottom_border(ht, row, col, solid_border)
```

Comment:
Should be able to do without the for loop I think?

a la:

```
> jams
                                                     Type                Price  
                                                     Strawberry           1.90  
                                                     Raspberry            2.10  
                                                     Plum                 1.80  

Column names: Type, Price

> bottom_border(jams)[matrix(c(FALSE, TRUE, TRUE, FALSE, FALSE, FALSE), ncol = 2)] <- .6
> jams
                                                     Type                Price  
                                                     Strawberry           1.90  
                                                   ───────────────              
                                                     Raspberry            2.10  
                                                   ───────────────              
                                                     Plum                 1.80  
                                                                 ───────────────

Column names: Type, Price

```

Reply:
Agreed. The earlier versions required access to neighbouring elements, so the loop made sense.